torch/csrc/jit/codegen/cuda/lower_unroll.cpp

```diff
@@ -32,6 +32,228 @@ kir::ForLoop* cloneLoopNest(const kir::ForLoop* for_loop, bool unroll = false) {
+}
+
+kir::Val* setupNamedScalar(
+    kir::Scope& body,
+    kir::Val* val,
+    const std::string& name) {
+  kir::IrBuilder ir_builder(GpuLower::current()->kernel());
+  kir::Int* one = ir_builder.create<kir::Int>(1);
+
+  auto namedScalar = ir_builder.namedSetExpr(name, val);
+  auto alloc =
+      ir_builder.create<kir::Allocate>(namedScalar, MemoryType::Local, one);
+  body.push_back(alloc);
+  body.push_back(namedScalar->definition());
+
+  return namedScalar;
+}
+
+kir::ForLoop* handleMisalignedVectorization(
+    std::vector<kir::ForLoop*> loop_structure,
+    const kir::ForLoop* for_loop) {
+  // body -> allocate, read, compute, write for-loops
+  kir::IrBuilder ir_builder(GpuLower::current()->kernel());
+  kir::Int* zero = ir_builder.create<kir::Int>(0);
+
+  // create new base For-Loop
+  const auto new_loop = ir_builder.create<kir::ForLoop>(
+      for_loop->index(), for_loop->extent(), for_loop->iter_domain());
+
+  // Find child For-Loops and add remaining expressions to base For-Loop
+  auto child_loops = parseVectorizedForLoop(for_loop, new_loop);
+
+  // Find the first vectorize set - either read or write
+  // Add child For-Loop to loop_structure
+  // Enable vectorize flag in child For-Loop
+  auto vec_expr = findVectorizedSet(loop_structure, child_loops);
+  TORCH_INTERNAL_ASSERT(vec_expr != nullptr);
+  TORCH_INTERNAL_ASSERT(vec_expr->outputs().front()->isA<kir::TensorView>());
+  TORCH_INTERNAL_ASSERT(vec_expr->inputs().front()->isA<kir::TensorView>());
+
+  // out_tv is the TensorView with the misaligned vec iterDomain
+  auto out_tv = vec_expr->outputs().front()->as<kir::TensorView>();
+  auto in_tv = vec_expr->inputs().front()->as<kir::TensorView>();
+  kir::TensorView* vec_tv =
+      (out_tv->memoryType() == MemoryType::Local) ? out_tv : in_tv;
+
+  const auto pred = PredicateCompute::getInlinePredicate(
+      vec_expr, loop_structure, nullptr, false, true);
+
+  kir::IfThenElse* pred_ite = ir_builder.create<kir::IfThenElse>(pred);
+  new_loop->body().push_back(pred_ite);
+
+  // Generate vectorize index
+  auto index = (out_tv->memoryType() == MemoryType::Global)
+      ? Index::getConsumerIndex(in_tv->fuserTv(), loop_structure)
+      : Index::getProducerIndex(
+            in_tv->fuserTv(), out_tv->fuserTv(), loop_structure);
+
+  //  Get full extent for merged root domains
+  auto extent = getVectorizeExtent(index);
+
+  auto vector_size =
+      vec_tv->domain()->domain().back()->extent()->as<kir::Int>();
+
+  // Disable vectorize flag in child For-Loop
+  loop_structure.back()->setVectorize(false);
+
+  auto base_address_val = generateBaseIndex(index);
+  auto base_address =
+      setupNamedScalar(pred_ite->thenBody(), base_address_val, "base_address");
+
+  auto a = ir_builder.ceilDivExpr(base_address, vector_size);
+  auto b = ir_builder.mulExpr(a, vector_size);
+  auto shift_val = ir_builder.subExpr(b, base_address);
+  auto shift = setupNamedScalar(pred_ite->thenBody(), shift_val, "shift");
```

Comment:
I'm still confused here. This is the corresponding part from the generated code of the MisalignedPointwise test:

```
      int64_t base_address;
      base_address = ((blockIdx.x * (1 * T1.size[1])) + (((ki70 * ((ceilDiv(256, 4)) * 4)) + ((threadIdx.x * 4) + 0)) * 1));
      int64_t shift;
      shift = (((ceilDiv(base_address, 4)) * 4) - base_address);
```

Isn't this, `(((ki70 * ((ceilDiv(256, 4)) * 4)) + ((threadIdx.x * 4) + 0)) * 1)`, always divisible by `vector_size`? If so, `base_address` is always divisible for `blockIdx.x * (1 * T1.size[1])` is divisible by `vector_size`. So, for example, thread blocks with `blockIdx.x == 0`, `base_address` is always divisible by `vector_size`, so `shift` is always zero for the thread blocks. Does this really address misaligned accesses? I wonder if the actual address of `vec_tv` needs to be considered.

Reply:
Base Address isn't always divisible be vector size.

If blockIdx.x == 1 and the tensor is [32, 47], is that address aligned.

It is 47 % 4 == 3.

If blockIdx.x == 2, then it is 94 % 4 == 2.

The first index is always aligned.

Reply:
Why didn't you consider a blockIdx.x > 0?